plugins/poll/lib/poll.rb

```diff
@@ -31,43 +41,100 @@ def self.vote(user, post_id, poll_name, options)
               obj << option.id if option.poll_votes.where(user_id: user.id).exists?
             end
 
-        # remove non-selected votes
-        PollVote.where(poll: poll, user: user).where.not(poll_option_id: new_option_ids).delete_all
+        if poll.type == "irv"
+          # for IRV, we need to remove all votes and re-create them as there is no way to update them due to lack of primary key.
+          PollVote.where(poll: poll, user: user).delete_all
+          creation_set = new_option_ids
+        else
+          # remove non-selected votes
+          PollVote
+            .where(poll: poll, user: user)
+            .where.not(poll_option_id: new_option_ids)
+            .delete_all
+          creation_set = new_option_ids - old_option_ids
+        end
 
         # create missing votes
-        (new_option_ids - old_option_ids).each do |option_id|
-          PollVote.create!(poll: poll, user: user, poll_option_id: option_id)
+        creation_set.each do |option_id|
+          if poll.type == "irv"
+            PollVote.create!(
+              poll: poll,
+              user: user,
+              poll_option_id: option_id,
+              rank:
+                options.find { |o| o[:digest] == poll.poll_options.find(option_id).digest }[:rank],
```

Comment:
Maybe extract `poll.poll_options.find(option_id).digest` into a variable to avoid doing the same work for each option?

Reply:
great suggestion, done!